### PR TITLE
safer bad egg battle engine

### PR DESCRIPTION
**Pull request by kleeenexfeu**
Prevents Get/SetBoxMonData from writing in the pokemon data substructure when checking if it"s a bad egg. It's safer, and takes less space. As we've seen in rh-hideout#1908 reading wrong data with these functions will corrupt different part of the RAM. It has already happened, might be happening right now, and will probably happen again.
With this change we avoid a slow corruption of the RAM.

It was already discussed in rh-hideout#1912 , but now I've been able to test it in a whole playthrough without any issue. Corrupted mon are treated as bad eggs, non corrupted mons are fin.
I insist that it prevents future (and maybe current) problems with the usage of this function that we may introduce.
There isn't a single instance of gamefreak misusing these functions, so if anything goes wrong it'll be because of our changes.
One more point I want to bring up since it was discussed in the discord, and now that I've looked more deeply in the code, I can confirm that it being included in the battle engine does make sense, because Get/SetBoxMonData are actually used many times there, often indirectly (and that's what made the fix in rh-hideout#1908 difficult for me), because there always a function that calls a function... that end up using either GetBoxMonData or SetBoxMonData. They are constantly used to fill the battle data, check which real species a pokémon is in battle (like when they are transformed, ditto isn't treated as the species it's transformed into, for instance) when they mega evolve, change form, when we recalculate their stats, when pokemon level up, evolve, learn a new move. rh-hideout#1908 shows that it's used in the AI too (because ofc the AI wants to check mons in its party). And that's only the ones I can think of.
I think that'd be even better to include it in the 3 main branches by default, because, directly or indirectly, they must use GetBoxMonData/SetBoxMonData

Note: This is essentially the same code for which I had done a pull request, it's been now tested with a longer playtime.

### CODE REVIEW

File: src/pokemon.c [hunk 1]
```diff
@@ -3424,13 +3435,13 @@ u32 GetBoxMonData(struct BoxPokemon *boxMon, s32 field, u8 *data)
...
     case MON_DATA_SANITY_IS_BAD_EGG:
-        retVal = boxMon->isBadEgg;
+        retVal = IsBoxMonBadEgg(boxMon, field);
         break;
     case MON_DATA_SANITY_HAS_SPECIES:
         retVal = boxMon->hasSpecies;
         break;
     case MON_DATA_SANITY_IS_EGG:
-        retVal = boxMon->isEgg;
+        retVal = boxMon->isEgg | IsBoxMonBadEgg(boxMon, field);
```
Comment (AsparagusEduardo) [inline src/pokemon.c:3444]: || instead of |

**Review by AsparagusEduardo — COMMENTED** [review]
This is out of scope for any of the 3 branches. It should be added to the pret pokeemerald wiki instead.

File: src/pokemon.c (hunk 1 above)
Comment (LOuroboros) [reply src/pokemon.c:3444]: I don't think that's correct? `||` is a logical OR used for comparisons. Here a bitwise OR is being used to decide if `retVal` should be set to either 0 or 1.

Comment (kleeenexfeu) [reply src/pokemon.c:3444]: It's intentional, either can be only 0 or 1, and if one of any of these are 1, retVal should be one.
So if the result of boxMon->isEgg is in r0 and the result of IsBoxMonBadEgg is in r1, I would do
orr r0, r1
And then return r0, which is what this part of the code does

**Comment by LOuroboros** [comment]
I still stand by my opinion that this change is outside of the scope for any of the 3 feature branches which deal with the game's battle engine, its item data and its species data.

A better approach would be to either write up a comment about this and then submit it as a PR to Pret's Pokeemerald repository, or to write an article about it in that repository's Wiki.

**Comment by kleeenexfeu** [comment]
> This is out of scope for any of the 3 branches. It should be added to the pret pokeemerald wiki instead.

It prevents bugs that **we** may introduce, it doesn't fix anything from pokeemerald.
Although I agree with the princple that it's a "bugfix", shouldn't be something that we take care of?

**Comment by LOuroboros** [comment]
> It prevents bugs that **we** may introduce, it doesn't fix anything from pokeemerald.

But it's still fundamentally outside of the scope for any of the 3 feature branches.

`GetBoxMonData` is a function that can return the parameters of a Pokémon stored in the savefile, which is unrelated to the data that conforms the battle engine, or the data that composes an item, or the data that composes a species in the game.

I said this before, but I honestly believe that the bugs that **we** introduce should be taken care of directly, that we should not be misusing functions, and that when we do, we should correct that.

**Comment by kleeenexfeu** [comment]
> > It prevents bugs that **we** may introduce, it doesn't fix anything from pokeemerald.
> 
> But it's still fundamentally outside of the scope for any of the 3 feature branches.
> 
> `GetBoxMonData` is a function that can return the parameters of a Pokémon stored in the savefile, which is unrelated to the data that conforms the battle engine, or the data that composes an item, or the data that composes a species in the game.
> 
> I said this before, but I honestly believe that the bugs that **we** introduce should be taken care of directly, that we should not be misusing functions, and that when we do, we should correct that.

I agree in theory.

The problem in practice with this kind of bug is that it may slowly corrupt your RAM, for weeks, and you won't notice it. It can also break your game immediately depending of the data it corrupts. It's random and unpredictable. So yeah, if we could just fix the cause of the bug right away, I wouldn't bother with this PR, but maybe there are many more of them present in the BE and we don't know yet about it.

File: src/pokemon.c (hunk 1 above)
Comment (AsparagusEduardo) [reply src/pokemon.c:3444]: My bad.

**Comment by LOuroboros** [comment]
> The problem in practice with this kind of bug is that it may slowly corrupt your RAM, for weeks, and you won't notice it. It can also break your game immediately depending of the data it corrupts. It's random and unpredictable.

If you brought this up in Pret, perhaps you could implement it in their repository directly then, not as a comment, but under the `BUGFIX` preproc config, which realistically, everyone should be enabling in their projects.

**Comment by LOuroboros** [comment]
So, I guess we can all agree this won't be merged because this modification is out of scope for any of the 3 branches, so I'm closing it.

I once again encourage you to write an article about this modification in Pret's Wiki though @kleeenexfeu. It looks legitimately useful.

**Pull request closed without merging**